docs/source/development/eep-02-typing.md

```diff
@@ -0,0 +1,1474 @@
+- The basic namespaces help to quickly see what is influenced by a specific option. This
+  works especially well to distinguish stopping options and convergence criteria from
+  other tuning parameters of the algorithms. However, it would be enough to keep them as
+  a naming convention if we find it hard to support the `.` notation.
+- All options are documented in the estimagic documentation, i.e. we do not link to the
+  docs of original packages.
+
+**Problems**
+
+- There is no autocomplete and the only way to find out which options are supported is
+  the documentation.
+- A small typo in an option name can easily lead to the option being discarded.
+- Option dictionaries can grow very big.
+- The fact that option dictionaries are mutable can lead to errors, for example when a
+  user wants to try out a grid of values for one tuning parameter while keeping all
+  other options constant.
+
+**Secondary problems**
+
+The following problems are not related to the specific goals of this enhancement
+proposal but it might be a good idea to address them in the same deprecation cycle.
+
+- In an effort to make everything very descriptive, some names got too long. For example
+  `"convergence.absolute_gradient_tolerance"` is very long but most people are so
+  familiar with reading `"gtol_abs"` (from SciPy and NLopt) that
+  `"convergence.gtol_abs"` would be a better name.
+- It might have been a bad idea to harmonize default values for similar options that
+  appear in multiple optimizers. Sometimes, the options, while similar in spirit, are
+  defined slightly differently and usually algorithm developers will set all tuning
+  parameters to maximize performance on a benchmark set they care about. If we change
+  how options are handled in estimagic, we should consider just harmonizing names and
+  not default values.
+
+#### Proposal
+
+In total, we want to offer four entry points for the configuration of optimizers:
+
+1. Instead of passing an `Algorithm` class (as described in
+   [Algorithm Selection](algorithm-selection)) the user can create an instance of their
+   selected algorithm. When creating the instance, they have autocompletion for all
+   options supported by the selected algorithm. `Algorithm`s are immutable.
+1. Given an instance of an `Algorithm`, a user can easily create a modified copy of that
+   instance by using the `with_option` method.
+1. We can provide additional methods `with_stopping` and `with_convergence` that call
+   `with_option` internally but provide two additional features:
+   1. They validate that the option is indeed a stopping/convergence criterion.
+   1. They allow to omit the `convergence_` or `stopping_` at the beginning of the
+      option name and can thus reduce repetition in the option names.
+1. As before, the user can pass a global set of options to `maximize` or `minimize`. We
+   continue to support option dictionaries but also allow `AlgorithmOption` objects that
```

Comment:
It is a very good idea to have this for backwards compatibility. However, I would not show this in the documentation and rather give the user a more biased explanation for configuration (in favor of configuring the algorithms directly).

Reply:
Agree and added a note.